File: Question/3_Longes_Substring_Without_Repeating_Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int ans = INT_MIN;
        int left = -1;
        int right = 0;
        unordered_map<char, int> map; //char loc
        while(right < s.size()) {
            if(map.count(s[right]) && map[s[right]] > left) {
                left = map[s[right]];
            }
            map[s[right]] = right;
            ans = max(ans, right - left);
            ++right;
        }
        return ans;
    }
};

File: Question_Java/3_Longes_Substring_Without_Repeating_Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> mem = new HashMap<>();
        int ans = 0;
        int left = 0;
        for (int i = 0; i < s.length(); ++i) {
            char ch = s.charAt(i);
            if (!mem.containsKey(ch) || mem.get(ch) < left) {
                ans = Math.max(ans, i - left + 1);
            } else {
                left = mem.get(ch) + 1;
            }
            mem.put(ch, i);
        }
        return ans;
    }
}